python/tools/recorded-call-graph-metrics/cg_trace.py

```diff
@@ -0,0 +1,224 @@
+import dataclasses
+import csv
+import xml.etree.ElementTree as ET
+
+# Copy-Paste and uncomment for interactive ipython sessions
+# import IPython; IPython.embed(); sys.exit()
+
+
+@dataclasses.dataclass(frozen=True)
+class Call():
+    """A call to a callable
+    """
+    filename: str
+    linenum: int
+    inst_index: int
+
+    @classmethod
+    def from_frame(cls, frame, debugger: bdb.Bdb):
+        code = frame.f_code
+
+        # Uncomment to see the bytecode
+        # b = dis.Bytecode(frame.f_code, current_offset=frame.f_lasti)
+        # print(b.dis(), file=sys.__stderr__)
+
+        return cls(
+            filename = debugger.canonic(code.co_filename),
+            linenum = frame.f_lineno,
+            inst_index = frame.f_lasti,
+        )
+
+
+@dataclasses.dataclass(frozen=True)
+class Callable():
+    """A callable (Function/Lambda) should (hopefully) be uniquely identified by its name and
+    location (filename+line number)
+
+    TODO: Callable is maybe not a good name, since classes with __call__ will return true
+          for the python code `callable(cls)` -- will have to consider how __call__ is handled
+    """
+    funcname: str
+    filename: str
+    linenum: int
+
+    @classmethod
+    def from_frame(cls, frame, debugger: bdb.Bdb):
+        code = frame.f_code
+        return cls(
+            funcname = code.co_name,
+            filename = debugger.canonic(code.co_filename),
+            linenum = frame.f_lineno,
+        )
+
+
+class CallGraphTracer(bdb.Bdb):
+    """Tracer that records calls being made
+
+    It would seem obvious that this should have extended `trace` library
+    (https://docs.python.org/3/library/trace.html), but that part is not extensible --
+    however, the basic debugger (bdb) is, and provides maybe a bit more help than just
+    using `sys.settrace` directly.
+    """
+
+    recorded_calls: set
+
+    def __init__(self):
+        self.recorded_calls = set()
+        super().__init__()
+
+    def user_call(self, frame, argument_list):
+        call = Call.from_frame(frame.f_back, self)
+        callable = Callable.from_frame(frame, self)
+
+        # _print(f'{call}  -> {callable}')
+        self.recorded_calls.add((call, callable))
+
+
+################################################################################
+# Export
+################################################################################
+
+
+class Exporter:
+
+    @staticmethod
+    def export(recorded_calls, outfile_path):
+        raise NotImplementedError()
+
+    @staticmethod
+    def dataclass_to_dict(obj):
+        d = dataclasses.asdict(obj)
+        prefix = obj.__class__.__name__.lower()
+        return {f"{prefix}_{key}": val for (key, val) in d.items()}
+
+
+class CSVExporter(Exporter):
+
+    @staticmethod
+    def export(recorded_calls, outfile_path):
+        with open(outfile_path, 'w', newline='') as csv_file:
+            writer = None
+            for (call, callable) in recorded_calls:
+
```

Comment:
This empty line is a bit strange. Was it introduced by an autoformatter? :thinking:

Reply:
nope. that was just me :)